server/src/main/java/io/crate/expression/scalar/conditional/MultiCaseFunction.java

```diff
@@ -0,0 +1,105 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.expression.scalar.conditional;
+
+import static io.crate.metadata.functions.TypeVariableConstraint.typeVariable;
+import static io.crate.types.TypeSignature.parseTypeSignature;
+
+import java.util.List;
+
+import io.crate.data.Input;
+import io.crate.expression.scalar.ScalarFunctionModule;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.BoundSignature;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+public class MultiCaseFunction extends Scalar<Object, Object> {
+
+    public static void register(ScalarFunctionModule module) {
+        // multi_case([boolean] conditions, [E] results) -> E
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.BOOLEAN_ARRAY.getTypeSignature(),
+                parseTypeSignature("array(E)"),
+                parseTypeSignature("E")
+            ).withTypeVariableConstraints(typeVariable("E")),
+            MultiCaseFunction::new
+        );
+        // multi_case([boolean] conditions, [E] results, E default) -> E
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.BOOLEAN_ARRAY.getTypeSignature(),
+                parseTypeSignature("array(E)"),
+                parseTypeSignature("E"),
+                parseTypeSignature("E")
+            ).withTypeVariableConstraints(typeVariable("E")),
+            MultiCaseFunction::new
+        );
+    }
+
+    public static final String NAME = "multi_case";
+
+    private final Signature signature;
+    private final BoundSignature boundSignature;
+
+    private MultiCaseFunction(Signature signature, BoundSignature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public BoundSignature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public Object evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<Object>[] args) {
+        List<Boolean> conditions = (List<Boolean>) args[0].value();
+        List<Object> values = null;
+        for (int i = 0; i < conditions.size(); i++) {
+            if (conditions.get(i)) {
+                if (values == null) {
+                    // only evaluate when there is a matching condition
+                    values = (List<Object>) args[1].value();
```

Comment:
If not mistaken, this way we have to evaluate all values when we encounter a matching condition.
This is problematic not only because of performance, but also take this example:
```
matriv=> create table test(a int);
CREATE TABLE
matriv=> insert into test(a) values(10);
INSERT 0 1
matriv=> select CASE  
   WHEN a > 10 THEN 3
   WHEN a > 5 THEN 2
   WHEN a > 2 THEN 1
   WHEN a < 0 THEN 1/ a - 10
END FROM test;
 case 
------
    2
(1 row)
```

How does this evaluate for CrateDB?

Reply:
```
cr>  create table test(a int);
CREATE OK, 1 row affected  (1.786 sec)
cr> insert into test(a) values(10);
INSERT OK, 1 row affected  (0.254 sec)
cr> select CASE
       WHEN a > 10 THEN 3
       WHEN a > 5 THEN 2
       WHEN a > 2 THEN 1
       WHEN a < 0 THEN 1/ a - 10
    END FROM test;
+------------------------------------------------------------------------------------------+
| multi_case(_array((a > 10), (a > 5), (a > 2), (a < 0)), _array(3, 2, 1, ((1 / a) - 10))) |
+------------------------------------------------------------------------------------------+
|                                                                                        2 |
+------------------------------------------------------------------------------------------+
SELECT 1 row in set (0.622 sec)
cr>
```

Reply:
It seems to evaluate correctly.

Reply:
Also psql evaluates the whole `CASE` Statement at once:

```
mkleen=# select CASE
       WHEN a > 10 THEN 3
       WHEN a > 5 THEN 2
       WHEN a > 2 THEN 1
       WHEN a < 0 THEN 1/0
    END FROM test;
ERROR:  division by zero
```
vs:

```
cr> select CASE
       WHEN a > 10 THEN 3
       WHEN a > 5 THEN 2
       WHEN a > 2 THEN 1
       WHEN a < 0 THEN 1/0
    END FROM test;
SQLParseException[/ by zero]
```

Reply:
Yep, my bad, we need paretheses:
```
matriv=> select CASE  
   WHEN a > 10 THEN 3
   WHEN a > 5 THEN 2
   WHEN a > 2 THEN 1
   WHEN a < 0 THEN 1/ (a - 10)
END FROM test;
 case 
------
    2
(1 row)
````
For CrateDB we get division by 0

Reply:
Fix is easy: https://github.com/crate/crate/pull/13401/commits/d5e97eab0d943a48c0052941bb3dc5f9977bc1a5 maybe we can do something more elegant though.